Prem/move migrations to dedicated process

prem-fissionhq opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [ ] Risk label is set correctly
- [ ] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [ ] QA labels are set correctly
- [ ] I am certain that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

gherceg committed:
resolve x_frame_options related warning

default to never allowing views to load inside of a frame, and adding a decorator to views that load inside a frame

dependabot[bot] committed:
Bump django-crispy-forms from 1.7.0 to 1.9.2 in /requirements

Bumps [django-crispy-forms](https://github.com/django-crispy-forms/django-crispy-forms) from 1.7.0 to 1.9.2.
- [Release notes](https://github.com/django-crispy-forms/django-crispy-forms/releases)
- [Changelog](https://github.com/django-crispy-forms/django-crispy-forms/blob/master/CHANGELOG.md)
- [Commits](https://github.com/django-crispy-forms/django-crispy-forms/compare/1.7.0...1.9.2)

Signed-off-by: dependabot[bot] <support@github.com>

millerdev committed:
Undo incorrect dependabot changes

proteusvacuum committed:
Only allow ipv4 addresses

snopoke committed:
drop old get_ledger_values_for_cases_2 sql function

gherceg committed:
add decorator to relogin view

esoergel committed:
Add ESQuery wrapper for user data searches

esoergel committed:
Write a basic test for filtering by user data

esoergel committed:
Make a note about the various field names

dependabot[bot] committed:
Bump email-validator from 1.0.3 to 1.1.1 in /requirements

Bumps [email-validator](https://github.com/JoshData/python-email-validator) from 1.0.3 to 1.1.1.
- [Release notes](https://github.com/JoshData/python-email-validator/releases)
- [Commits](https://github.com/JoshData/python-email-validator/compare/v1.0.3...v1.1.1)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump setproctitle from 1.1.9 to 1.1.10 in /requirements

Bumps [setproctitle](https://github.com/dvarrazzo/py-setproctitle) from 1.1.9 to 1.1.10.
- [Release notes](https://github.com/dvarrazzo/py-setproctitle/releases)
- [Changelog](https://github.com/dvarrazzo/py-setproctitle/blob/master/HISTORY.rst)
- [Commits](https://github.com/dvarrazzo/py-setproctitle/compare/version-1.1.9...version-1.1.10)

Signed-off-by: dependabot[bot] <support@github.com>

gherceg committed:
add message subevent link to inbound and outbound survey sms events

gherceg committed:
fix fetching pk on type None

mkangia committed:
remove redundant icds_nic

mkangia committed:
fix empty collection

mkangia committed:
fix test

mkangia committed:
move skip phone entry to settings

mkangia committed:
stickler

mkangia committed:
fix test for real

mkangia committed:
allow ignoring demo user submissions

mkangia committed:
placed correctly before localsettings import

mkangia committed:
update occurrences

dependabot[bot] committed:
Bump graphviz from 0.13.1 to 0.14.2 in /requirements

Bumps [graphviz](https://github.com/xflr6/graphviz) from 0.13.1 to 0.14.2.
- [Release notes](https://github.com/xflr6/graphviz/releases)
- [Changelog](https://github.com/xflr6/graphviz/blob/master/CHANGES.txt)
- [Commits](https://github.com/xflr6/graphviz/compare/0.13.1...0.14.2)

Signed-off-by: dependabot[bot] <support@github.com>

gherceg committed:
add display name to gateway form

cleaned up flake warnings too

gherceg committed:
Update staging.yaml

gherceg committed:
Update staging.yaml

gherceg committed:
Update staging.yaml

gherceg committed:
Update staging.yaml

proteusvacuum committed:
Allow adding and removing of translations

gherceg committed:
only fetch subevent id

mjriley committed:
Updated data table formatting to fill the width available

dependabot[bot] committed:
Bump cryptography from 3.1.1 to 3.2 in /requirements

Bumps [cryptography](https://github.com/pyca/cryptography) from 3.1.1 to 3.2.
- [Release notes](https://github.com/pyca/cryptography/releases)
- [Changelog](https://github.com/pyca/cryptography/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pyca/cryptography/compare/3.1.1...3.2)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump datadog from 0.15.0 to 0.39.0 in /requirements

Bumps [datadog](https://github.com/DataDog/datadogpy) from 0.15.0 to 0.39.0.
- [Release notes](https://github.com/DataDog/datadogpy/releases)
- [Changelog](https://github.com/DataDog/datadogpy/blob/master/CHANGELOG.md)
- [Commits](https://github.com/DataDog/datadogpy/compare/v0.15.0...v0.39.0)

Signed-off-by: dependabot[bot] <support@github.com>

snopoke committed:
add config for label-bot

snopoke committed:
don't set to None

snopoke committed:
mark ES mapping changes as reindex/migration

mkangia committed:
remove model

snopoke committed:
disable unused actions

snopoke committed:
add link to docs

snopoke committed:
update label bot to handle WIP workflow

snopoke committed:
add the labels we use

snopoke committed:
add public docs about our usage of labels

stephherbers committed:
now displays password bullets when auth_token is entered

mkangia committed:
fix test patch for removed method

snopoke committed:
add note about external contributors

esoergel committed:
reuse es instance

Co-authored-by: Farid Rener <proteusvacuum@users.noreply.github.com>

gherceg committed:
remove display name validation

esoergel committed:
Flesh out tests a bit more

orangejenny committed:
Update staging.yaml

stephherbers committed:
'Actions' column not visible for 'can view' users

stephherbers committed:
minor-- fixed indentation

mkangia committed:
track if user created

mkangia committed:
fix failing test

orangejenny committed:
Fixed web apps tile display for medium-width screens

mkangia committed:
add missing test

biyeun committed:
Update translations.

[ci skip]

gherceg committed:
add optional arguments to bootstrap gateway command

orangejenny committed:
Updated get_linked_domains to exclude deleted domains

snopoke committed:
fix spelling

snopoke committed:
remove unused code

orangejenny committed:
Added predicted risk labels

orangejenny committed:
Added git-heat script [ci skip]

orangejenny committed:
Switched from 'Expected Risk' to 'Risk'

mjriley committed:
fixed resource leak with GzipFile

dannyroberts committed:
WIP from oct 29 danny prem work session

esoergel committed:
Use simpler blob name

With multiple extends, this can easily pass the filename character limit

esoergel committed:
Add some logging to show progress through blobs

orangejenny committed:
Removed function wrappers for shell scripts

orangejenny committed:
Updated scripts/ files to be executable

proteusvacuum committed:
Add click tracking for app diff buttons on releases table

gherceg committed:
troubleshoot testing woes

orangejenny committed:
Update CONTRIBUTING.rst

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

orangejenny committed:
Added support for deleting linked reports

proteusvacuum committed:
Move track.event out of html

orangejenny committed:
Fixed YAML in risk labels config

orangejenny committed:
Added README for corehq.apps.users [ci skip]

esoergel committed:
Restructure blob exporter to accept list of blobs to skip

gherceg committed:
add test that inbound sms is linked to subevent

orangejenny committed:
Update corehq/apps/users/README.rst

Co-authored-by: Steph <36681924+stephherbers@users.noreply.github.com>

esoergel committed:
Skip blobs before fetching them

orangejenny committed:
Updated to avoid needing to translate

orangejenny committed:
Update corehq/apps/users/README.rst

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

orangejenny committed:
Added more about terminology [ci skip]

orangejenny committed:
Fixed typo [ci skip]

orangejenny committed:
Fixed typo [ci skip]

sravfeyn committed:
Use keywords in ES aggs

sravfeyn committed:
Revert "Use keywords in ES aggs"

This reverts commit 0578f030e40af97199b5715356ca2ade0aac682f.

orangejenny committed:
Added docs on initial page data in js tests

orangejenny committed:
Reverted modes for non-executable files

orangejenny committed:
Reverted modes for python files

esoergel committed:
Pull already_exported to where it's used

I think this makes sense, since it's not really an attribute of the
archive anyways.

gherceg committed:
add test for no link and cleanup

gherceg committed:
use method on SQLXFormsSession to fetch subevent

gherceg committed:
remove unnecessary import

gherceg committed:
add test for multiple subevents

orangejenny committed:
Update staging.yaml

orangejenny committed:
Update staging.yaml

esoergel committed:
Update CODEOWNERS

dependabot[bot] committed:
Bump django-simple-captcha from 0.5.12 to 0.5.13 in /requirements

Bumps [django-simple-captcha](https://github.com/mbi/django-simple-captcha) from 0.5.12 to 0.5.13.
- [Release notes](https://github.com/mbi/django-simple-captcha/releases)
- [Changelog](https://github.com/mbi/django-simple-captcha/blob/master/CHANGES)
- [Commits](https://github.com/mbi/django-simple-captcha/compare/v0.5.12...v0.5.13)

Signed-off-by: dependabot[bot] <support@github.com>

snopoke committed:
drop grapevine rate limit

SMS callbacks are being hitting the rate limit every day

gherceg committed:
tear down properly and use contextlib

millerdev committed:
updating files: scripts/staging.yaml

[ci skip]

millerdev committed:
updating files: scripts/staging.yaml

[ci skip]

gherceg committed:
unique domain name

kaapstorm committed:
simple_post to notify on exceptions & HTTP errors

kaapstorm committed:
Add test

orangejenny committed:
Added a few high-risk files

gherceg committed:
Update staging.yaml

kaapstorm committed:
Don't send empty message if response text is empty

snopoke committed:
remove forms_not_in_es command

There is a better command to use which is `stale_data_in_es`.
It is better because it is resumable, has progress output, can
operate on forms and cases, can be filtered by date range.

snopoke committed:
allow restricting stale_data_in_es to either couch or sql

snopoke committed:
refactor & rename

snopoke committed:
less verbose output from backfill commands

dependabot[bot] committed:
Bump werkzeug from 0.15.3 to 1.0.1 in /requirements

Bumps [werkzeug](https://github.com/pallets/werkzeug) from 0.15.3 to 1.0.1.
- [Release notes](https://github.com/pallets/werkzeug/releases)
- [Changelog](https://github.com/pallets/werkzeug/blob/master/CHANGES.rst)
- [Commits](https://github.com/pallets/werkzeug/compare/0.15.3...1.0.1)

Signed-off-by: dependabot[bot] <support@github.com>

kaapstorm committed:
'updating files: scripts/staging.yaml'

'[ci skip]'

biyeun committed:
Update translations.

[ci skip]

biyeun committed:
Revert "drop old get_ledger_values_for_cases_2 sql function"

calellowitz committed:
Revert "Fixed web apps tile display for medium-width screens"

orangejenny committed:
Fixed Web Apps tile misalignment

ShivamPokhriyal committed:
Re-add translations for CommCare 2.50

snopoke committed:
use class methods

dannyroberts committed:
Add more detailed assurance steps to PR template

snopoke committed:
fix test mock references

orangejenny committed:
Update staging.yaml

dannyroberts committed:
Rearrange Safety Assurance section of PR template

orangejenny committed:
Added query support to case search endpoint

mjriley committed:
Update staging.yaml

orangejenny committed:
Renamed _query to _xpath_query

orangejenny committed:
Update staging.yaml

esoergel committed:
Always be verbose

This isn't actually that verbose, and not having logging is extremely disconcerting

esoergel committed:
Minor rearranging to make this easier to read

esoergel committed:
Add progress bar to loaders

sravfeyn committed:
Add ids_query in prep for ILM

stephherbers committed:
Update staging.yaml

dependabot[bot] committed:
Bump django-braces from 1.13.0 to 1.14.0 in /requirements

Bumps [django-braces](https://github.com/brack3t/django-braces) from 1.13.0 to 1.14.0.
- [Release notes](https://github.com/brack3t/django-braces/releases)
- [Changelog](https://github.com/brack3t/django-braces/blob/master/docs/changelog.rst)
- [Commits](https://github.com/brack3t/django-braces/compare/v1.13.0...v1.14.0)

Signed-off-by: dependabot[bot] <support@github.com>

orangejenny committed:
Update staging.yaml

orangejenny committed:
Added notify_exception for users created with user_data kwarg

orangejenny committed:
Switched user_data kwarg for metadata in TempCommCareUser.__init__

mjriley committed:
Update staging.yaml

orangejenny committed:
Minor updates to js tests docs

millerdev committed:
updating files: scripts/staging.yaml

[ci skip]

dependabot[bot] committed:
Bump psutil from 5.7.2 to 5.7.3 in /requirements

Bumps [psutil](https://github.com/giampaolo/psutil) from 5.7.2 to 5.7.3.
- [Release notes](https://github.com/giampaolo/psutil/releases)
- [Changelog](https://github.com/giampaolo/psutil/blob/master/HISTORY.rst)
- [Commits](https://github.com/giampaolo/psutil/compare/release-5.7.2...release-5.7.3)

Signed-off-by: dependabot[bot] <support@github.com>

gherceg committed:
Update staging.yaml

biyeun committed:
add toggle for listing active couch domains

biyeun committed:
add utilities for syncing stale es data with active couch domains

biyeun committed:
add task for periodically syncing stale es data in active couch domains

biyeun committed:
add management command for syncing stale es data in high priority couch domains

biyeun committed:
update metric to return stale doc count in ES

biyeun committed:
lint

biyeun committed:
fix cyclic import error

Commit:
Update staging.yml

Commit:
to run migrations only on a process_num when process_num=0 and dedicated_migration_process

Commit:
to run migrations only on a process_num when process_num=0 and dedicated_migration_process

prem-fissionhq committed:
Merge branch 'master' into prem/move_migrations_to_dedicated_process

Commit:
removed print statements

Commit:
removed print statements

prem-fissionhq committed:
removed print statement

prem-fissionhq committed:
removed print statement

dannyroberts commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -106,7 +107,13 @@ def run(self):
         with configure_scope() as scope:
             scope.set_tag("pillow_name", self.get_name())
-        self.process_changes(since=self.get_last_checkpoint_sequence(), forever=True)
+        if self.process_num==0  and self.is_dedicated_migration_process:
```
nit: stickler is not going to like this spacing

prem-fissionhq replied:
Noted Danny. thanks

dannyroberts commented on corehq/pillows/application.py:
```diff
@@ -35,16 +35,18 @@ def get_app_to_elasticsearch_pillow(pillow_id='ApplicationToElasticsearchPillow'
     )
     change_feed = KafkaChangeFeed(
-        topics=[topics.APP], client_id='apps-to-es', num_processes=num_processes, process_num=process_num
+        topics=[topics.APP], client_id='apps-to-es', num_processes=num_processes, process_num=process_num, dedicated_migration_process=dedicated_migration_process,
```
nit: stickler is not going to like the long line

prem-fissionhq replied:
Noted Danny. thanks

dannyroberts reviewed (commented):
I think this logic (`is_dedicated_migration_process=dedicated_migration_process and (process_num == 0)`) only needs to be added wherever you see `run_migrations=(process_num == 0)`, because other pillows don't actually have any migrations to run, and so it would be weird for them to support running migrations in a side process. Does that make sense?

prem-fissionhq commented:
> I think this logic (`is_dedicated_migration_process=dedicated_migration_process and (process_num == 0)`) only needs to be added wherever you see `run_migrations=(process_num == 0)`, because other pillows don't actually have any migrations to run, and so it would be weird for them to support running migrations in a side process. Does that make sense?
yes, @dannyroberts . will remove them from other places. Apart from that do you have any other suggestions on one those have run_migrations? thanks

Commit:
removed dedicated_migration process configurations from the pillows where it is not necessary

Commit:
removed dedicated_migration process configurations from the pillows where it is not necessary

Commit:
Merge branch 'prem/move_migrations_to_dedicated_process' of https://github.com/prem-fissionhq/commcare-hq into prem/move_migrations_to_dedicated_process

prem-fissionhq commented:
@dannyroberts pushed my changes could you please review it. thanks.

Commit:
bug: missed this variable while copying. Passing process_num variable to the ConstructedPillow

dannyroberts commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -106,7 +107,13 @@ def run(self):
         with configure_scope() as scope:
             scope.set_tag("pillow_name", self.get_name())
-        self.process_changes(since=self.get_last_checkpoint_sequence(), forever=True)
+        if self.process_num == 0  and self.is_dedicated_migration_process:
```
I think this is redundant since it's always set to `is_dedicated_migration_process=dedicated_migration_process and (process_num == 0)`, right? I think it'd be less confusing if this line was just

```suggestion
        if self.is_dedicated_migration_process:
```
unless I'm missing something

prem-fissionhq replied:
No, Danny, you're correct. In order to capture both process_num and dedicated_migration_process, I placed it in that way. We can remove it

dannyroberts commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -429,6 +436,8 @@ def __init__(self, name, checkpoint, change_feed, processor,
 
         self._change_processed_event_handler = change_processed_event_handler
+        self.is_dedicated_migration_process = is_dedicated_migration_process
+        self.process_num = process_num
```
Where is `process_num` used? If it's just in the place where I commented above then I think you can remove it entirely

dannyroberts commented:
I think my only other comment is that there are a lot of minor but unrelated changes in this PR: blank lines that were added or removed, function parameters reordered without otherwise changing the function, etc. To clean it up I would usually do the following process

1. `git rebase master -i` and then mark all but the first commit with `f` for fixup
2. `git reset head^` to put you back on master with the changes intact but uncommitted
3. `git add -p .` to interactively add sections of code. In this step only add changes that are central to the PR
4. `git commit` with the usual nice message to summarize the change

This should give you a nice small PR that's simple to follow, and easier to test and reason with.

prem-fissionhq commented:
Sure Danny. Will update and let you know. thanks

Commit:
moving migrations to a dedicated_process_num (i.e on first process_num) and not to run any other processes when run_migrations is set to True.
To handle this behavior, introduced variable dedicated_migration_process which will be passed from the app_processes.yml if required.

Commit:
Merge branch 'prem/move_migrations_to_dedicated_process' of https://github.com/prem-fissionhq/commcare-hq into prem/move_migrations_to_dedicated_process

Commit:
moving migrations to a dedicated_process_num (i.e on first process_num) and not to run any other processes when run_migrations is set to True.
To handle this behavior, introduced variable dedicated_migration_process which will be passed from the app_processes.yml if required.

prem-fissionhq commented:
@dannyroberts as suggested, I followed the steps and pushed the changes. I still see a lot of files in the commit, will raise another PR
